data/src/dcc.rs

```diff
@@ -66,12 +66,37 @@ impl Send {
         }
     }
 
-    fn decode<'a>(mut args: impl Iterator<Item = &'a str>) -> Option<Self> {
-        let filename = args.next()?.to_string();
-        let host = args.next().and_then(decode_host)?;
-        let port = NonZeroU16::new(args.next()?.parse().ok()?);
-        let size = args.next()?.parse().ok()?;
-        let token = args.next();
+    fn decode<'a>(args: impl Iterator<Item = &'a str>) -> Option<Self> {
+        let mut args: Vec<&str> = args.collect();
+        args.reverse();
+
+        // if token doesn't exist
+        let mut token = None;
+        let mut size = args[0];
+        let mut port = args[1];
+        let mut host = args[2];
+        let mut filename = args.iter().skip(3);
+
+        // If token exists, port == 0
+        // args[1] == port, if token doesn't exists.
+        // args[2] == port, if token exists.
+        if args[1].parse::<u16>() == Ok(0) || args[2].parse::<u16>() == Ok(0) {
+            token = Some(args[0].parse::<NonZeroU16>().ok()?);
+            size = args[1];
+            port = args[2];
+            host = args[3];
+            filename = args.iter().skip(4);
```

Comment:
The logic doesn't seem right. If args[1] == 0 then we don't have token, but then you parse args[0] as token instead of size.

Having port 0 without token shouldn't be allowed, we may need to return None in that case

Reply:
the problem is that there is not a easy way to determine if token exists, so I don't the exact position of port.
This could be easier if instead of an iterator of strings. it was already a struct or if the token was marked as None if doesn't exists.
I am not familiar with irc or dcc, is there a reason why args is a iterator of strings or the position of the values?

Reply:
I think we can just check `args.len > 4 and args[len - 3] == 0` to be positive the last arg is a token w/ port 0 provided and enough args to include all the other required parameters.

Reply:
With that logic:

args = "my file name" 1402301083 12350 1453953495
args.len = 6
args[len-3] == 1402301083 == host
Here you are checking the host not the port

args = "my file name" 1402301083 12350 1453953495 token
args.len = 7
args[len-3] == 12350 == port
Here you are checking the port

not much different than this:
port.parse::<u16>() == Ok(0) || host.parse::<u16>() == Ok(0)

Reply:
What is the structure of token? can be modified to put a None in the String when it doesn't exist? 
This will make thinks easier.

Reply:
I forgot that port isn't always 0 when token is provided, for example the person responding to the reverse request will put their host / port in along with the token.

Thinking through this again, I think the most robust way is to iterate in reverse until we find the arg that successfully parses as IpAddr. Since this won't be token, size or port, this should always be reliable. We can then check if 2 or 3 args exist AFTER that arg to know 100% if token exists. And everything BEFORE that arg must be the filename.

Reply:
The problem is that IpAdrr just check if the string can be parsed to an u32.
```rust
fn decode_host(host: &str) -> Option<IpAddr> {
    match host.parse::<u32>() {
        Ok(n) => Some(IpAddr::V4(Ipv4Addr::from(n))),
        Err(_) => host.parse().ok(),
    }
}
```
So you can actually parse `size` and `port` as `IpAdrr`. Also if the `filename` include only numbers in the last word, it will be parsed as `IpAdrr`.

Reply:
Ahh right, ipv4 in this "spec" is sent as u32 BE encoding for the 4 bytes. Bummer! 

We could assume the _last_ successful IpAddr parsed with decode_host in rev iteration?

Reply:
@ameknite Checkout a46eec9 and let me know what you think.

The only potential gap I can think of is if token isn't set and filename has spaces + ends in an integer. But I think this is an extreme edge case that shouldn't be accounted for, as it's probably impossible to account for. So I think this is sufficient.